README.md
# Ueberauth HPID

> HP ID OAuth2 strategy for Überauth.

## Installation

1. Add `:ueberauth_hpid` to your list of dependencies in `mix.exs`:

    ```elixir
    def deps do
      [{:ueberauth_hpid, "~> 1.0"}]
    end
    ```

1. Add the strategy to your applications:

    ```elixir
    def application do
      [applications: [:ueberauth_hpid]]
    end
    ```

1. Add HPID to your Überauth configuration:

    ```elixir
    config :ueberauth, Ueberauth,
      providers: [
        hpid: {Ueberauth.Strategy.HPID, []}
      ]
    ```

1.  Update your provider configuration:

    ```elixir
    config :ueberauth, Ueberauth.Strategy.HPID.OAuth,
      client_id: System.get_env("HPID_CLIENT_ID"),
      client_secret: System.get_env("HPID_CLIENT_SECRET")
    ```

1.  Include the Überauth plug in your controller:

    ```elixir
    defmodule MyApp.AuthController do
      use MyApp.Web, :controller

      pipeline :browser do
        plug Ueberauth
        ...
       end
    end
    ```

1.  Create the request and callback routes if you haven't already:

    ```elixir
    scope "/auth", MyApp do
      pipe_through :browser

      get "/:provider", AuthController, :request
      get "/:provider/callback", AuthController, :callback
    end
    ```

1. Your controller needs to implement callbacks to deal with `Ueberauth.Auth` and `Ueberauth.Failure` responses.

For an example implementation see the [Überauth Example](https://github.com/ueberauth/ueberauth_example) application.

## Calling

Depending on the configured url you can initiate the request through:

    /auth/hpid

Or with options:

    /auth/hpid?scope=openid+profile+email
    
Or if using token workflow:

    /auth/hpid/callback?token=xyz

By default the requested scope is "openid+profile+email". This provides both read access to the HPID user profile. See more at [HPID's OAuth Documentation](https://developers.hp.com/hp-id/doc/authn-scopes). Scope can be configured either explicitly as a `scope` query value on the request path or in your configuration:

```elixir
config :ueberauth, Ueberauth,
  providers: [
    hpid: {Ueberauth.Strategy.HPID, [
      default_scope: "openid+profile+email",
      redirect_uri: "https://yourdomain.com"
    ]}
  ]
```

If you need to override the callback url (redirect_uri) at runtime, you can use the 
configuration in the Ueberauth.Strategy.HPID.OAuth:

```elixir
    config :ueberauth, Ueberauth.Strategy.HPID.OAuth,
      client_id: System.get_env("HPID_CLIENT_ID"),
      client_secret: System.get_env("HPID_CLIENT_SECRET"),
      redirect_uri: "https://yourdomain.com/auth/hpid/callback"
```

## License

Please see [LICENSE](https://github.com/VIPAAR/ueberauth_hpid/blob/master/LICENSE) for licensing details.



ChangeLog
1.0.4
==========
2019-01-02  Marcus Dillavou  <marcus.dillavou@helplightning.com>
	* Allow passing through the prompt variable to the auth route

1.0.3
==========
2018-12-21  Marcus Dillavou  <marcus.dillavou@helplightning.com>
	* Fix the default scope from being uri encoded

1.0.2
==========
2018-12-20  Marcus Dillavou  <marcus.dillavou@helplightning.com>
	* Added an optional redirect_uri parameter to override the callback url

1.0.1
==========
2018-11-05  Marcus Dillavou  <marcus.dillavou@helplightning.com>
	* Fixed ex_doc version

1.0.0
==========
2018-11-05  Marcus Dillavou  <marcus.dillavou@helplightning.com>
	* Initial Implementation



test/ueberauth_hpid_test.exs
defmodule UeberauthHPIDTest do
  use ExUnit.Case
  doctest UeberauthHPID
end


mix.exs
defmodule UeberauthHPID.MixProject do
  use Mix.Project

  @version "1.0.4"

  def project do
    [
      app: :ueberauth_hpid,
      version: @version,
      name: "Ueberauth HPID",
      package: package(),
      elixir: "~> 1.5",
      start_permanent: Mix.env() == :prod,
      source_url: "https://github.com/VIPAAR/ueberauth_hpid",
      homepage_url: "https://github.com/VIPAAR/ueberauth_hpid",
      description: description(),
      deps: deps(),
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger, :ueberauth, :oauth2]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:oauth2, "~> 0.9"},
      {:ueberauth, "~> 0.4"},

      # dev/test only dependencies
      {:credo, "~> 0.8", only: [:dev, :test]},

      # docs dependencies
      {:earmark, ">= 0.0.0", only: :dev},
      {:ex_doc, "~> 0.18.0", only: :dev}
    ]
  end

  defp docs do
    [extras: ["README.md"]]
  end

  defp description do
    "An Ueberauth strategy for using HP ID to authenticate your users."
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README.md", "LICENSE"],
      maintainers: ["Marcus Dillavou"],
      licenses: ["MIT"],
      links: %{GitHub: "https://github.com/VIPAAR/ueberauth_hpid"}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :ueberauth_hpid, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:ueberauth_hpid, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ueberauth/strategy/hpid/oauth.ex
defmodule Ueberauth.Strategy.HPID.OAuth do
  @moduledoc """
  An implementation of OAuth2 for HP ID.

  To add your `client_id` and `client_secret` include these values in your configuration.

      config :ueberauth, Ueberauth.Strategy.HPID.OAuth,
        client_id: System.get_env("HPID_CLIENT_ID"),
        client_secret: System.get_env("HPID_CLIENT_SECRET")
  """
  use OAuth2.Strategy

  # production host
  @prod_host "https://directory.id.hp.com"
  # staging host
  @stage_host "https://directory.stg.cd.id.hp.com"

  @doc """
  Construct a client for requests to HP ID.

  Optionally include any OAuth2 options here to be merged with the defaults.

      Ueberauth.Strategy.HPID.OAuth.client(redirect_uri: "http://localhost:4000/auth/hpid/callback")

  This will be setup automatically for you in `Ueberauth.Strategy.HPID`.
  These options are only useful for usage outside the normal callback phase of Ueberauth.
  """
  def client(opts \\ []) do
    config =
      :ueberauth
      |> Application.fetch_env!(Ueberauth.Strategy.HPID.OAuth)
      |> check_config_key_exists(:client_id)
      |> check_config_key_exists(:client_secret)

    client_opts =
      defaults()
      |> Keyword.merge(config)
      |> Keyword.merge(opts)

    OAuth2.Client.new(client_opts)
  end

  @doc """
  Provides the authorize url for the request phase of Ueberauth. No need to call this usually.
  """
  def authorize_url!(params \\ [], opts \\ []) do
    opts
    |> client
    |> OAuth2.Client.authorize_url!(params)
  end

  def get(token, url, headers \\ [], opts \\ []) do
    [token: token]
    |> client
    |> put_param("client_secret", client().client_secret)
    |> OAuth2.Client.get(url, headers, opts)
  end

  def post(token, url, body \\ "", headers \\ [], opts \\ []) do
    [token: token]
    |> client
    |> put_param("client_secret", client().client_secret)
    |> OAuth2.Client.post(url, body, headers, opts)
  end

  def get_token!(params \\ [], options \\ []) do
    headers = Keyword.get(options, :headers, [])
    options = Keyword.get(options, :options, [])
    client_options = Keyword.get(options, :client_options, [])
    client = OAuth2.Client.get_token!(client(client_options), params, headers, options)
    client.token
  end

  @doc ~S"""
  Validate a token, by matching the active status
  and client_id.
  """
  @spec validate(String.t()) :: boolean()
  def validate(token) do
    client_id =
      :ueberauth
      |> Application.fetch_env!(Ueberauth.Strategy.HPID.OAuth)
      |> Keyword.get(:client_id)

    # hpid requires very specific headers for this method!
    headers = [
      {"content-type", "application/x-www-form-urlencoded"},
      {"accept", "application/json"}
    ]

    # send the token in the body
    body = %{token: token.access_token}

    case __MODULE__.post(token, "/directory/v1/oauth/validate", body, headers) do
      {:ok, %OAuth2.Response{status_code: 401, body: _body}} ->
        false

      {:ok, %OAuth2.Response{status_code: status_code, body: user}}
      when status_code in 200..399 ->
        case user do
          %{"active" => true, "client_id" => ^client_id} -> true
          _ -> false
        end

      {:error, %OAuth2.Error{reason: _}} ->
        false
    end
  end

  # Strategy Callbacks

  def authorize_url(client, params) do
    OAuth2.Strategy.AuthCode.authorize_url(client, params)
  end

  def get_token(client, params, headers) do
    client
    |> put_param("client_secret", client.client_secret)
    |> put_header("Accept", "application/json")
    |> OAuth2.Strategy.AuthCode.get_token(params, headers)
  end

  def defaults do
    site_host = host()

    [
      strategy: __MODULE__,
      site: site_host,
      authorize_url: "#{site_host}/directory/v1/oauth/authorize",
      token_url: "#{site_host}/directory/v1/oauth/token"
    ]
  end

  defp check_config_key_exists(config, key) when is_list(config) do
    unless Keyword.has_key?(config, key) do
      raise "#{inspect(key)} missing from config :ueberauth, Ueberauth.Strategy.HPID"
    end

    config
  end

  defp check_config_key_exists(_, _) do
    raise "Config :ueberauth, Ueberauth.Strategy.HPID is not a keyword list, as expected"
  end

  defp host do
    case config()[:use_stage] do
      true -> @stage_host
      _ -> @prod_host
    end
  end

  defp config do
    Application.fetch_env!(:ueberauth, Ueberauth.Strategy.HPID.OAuth)
  end
end


lib/ueberauth/strategy/hpid.ex
defmodule Ueberauth.Strategy.HPID do
  @moduledoc """
  Provides an Ueberauth strategy for authenticating with HP ID.

  ### Setup

  Include the provider in your configuration for Ueberauth

      config :ueberauth, Ueberauth,
        providers: [
          hpid: { Ueberauth.Strategy.HPID, [] }
        ]

  Then include the configuration for HP ID.

      config :ueberauth, Ueberauth.Strategy.HPID.OAuth,
        client_id: System.get_env("HPID_CLIENT_ID"),
        client_secret: System.get_env("HPID_CLIENT_SECRET"),
        use_stage: True,
        redirect_uri: "https://example.com/auth/hpid/callback"

  The use_stage and redirect_uri are optional configurations:
      - use_stage :: boolean() ->  Enable / Disable the HP ID Staging server. Defaults to false
      - redirect_uri :: String.t() -> Allows overriding the callback url. This is useful for runtime configuration.


  If you haven't already, create a pipeline and setup routes for your callback handler

      pipeline :auth do
        Ueberauth.plug "/auth"
      end

      scope "/auth" do
        pipe_through [:browser, :auth]

        get "/:provider/callback", AuthController, :callback
      end


  Create an endpoint for the callback where you will handle the `Ueberauth.Auth` struct

      defmodule MyApp.AuthController do
        use MyApp.Web, :controller

        def callback_phase(%{ assigns: %{ ueberauth_failure: fails } } = conn, _params) do
          # do things with the failure
        end

        def callback_phase(%{ assigns: %{ ueberauth_auth: auth } } = conn, params) do
          # do things with the auth
        end
      end

  You can edit the behaviour of the Strategy by including some options when you register your provider.

  To set the default 'scopes' (permissions):

      config :ueberauth, Ueberauth,
        providers: [
          hpid: { Ueberauth.Strategy.HPID, [default_scope: "openid+profile+email"] }
        ]

  Default is empty ("openid+profile+email") which "Grants read-only access to profile information"
  """
  use Ueberauth.Strategy,
    default_scope: "openid profile email",
    oauth2_module: Ueberauth.Strategy.HPID.OAuth

  alias Ueberauth.Auth.Info
  alias Ueberauth.Auth.Credentials
  alias Ueberauth.Auth.Extra

  @doc """
  Handles the initial redirect to the HP ID authentication page.

  To customize the scope (permissions) that are requested by hpid include them as part of your url:

      "/auth/hpid?scope=openid+profile+email"

  You can also include a `state` param that hpid will return to you.
  """
  def handle_request!(conn) do
    scopes = conn.params["scope"] || option(conn, :default_scope)
    send_redirect_uri = Keyword.get(options(conn), :send_redirect_uri, true)

    opts =
      if send_redirect_uri do
        [redirect_uri: redirect_uri(conn), scope: scopes]
      else
        [scope: scopes]
      end

    # List of query parameters that we pass through IF
    #  clients sent them.
    # For more info, see:
    #  https://developers.hp.com/hp-id/doc/authn-grants#authorization-code-flow
    allowed_parameters = [:state, :prompt]

    opts =
      Enum.reduce(allowed_parameters, opts, fn(p, acc) ->
        case conn.params[Atom.to_string(p)] do
          nil -> acc
          v -> Keyword.put(acc, p, v)
        end
      end)

    module = option(conn, :oauth2_module)
    redirect!(conn, apply(module, :authorize_url!, [opts]))
  end

  @doc """
  Handles the callback from HP ID. When there is a failure from HP ID the failure is included in the
  `ueberauth_failure` struct. Otherwise the information returned from HP ID is returned in the `Ueberauth.Auth` struct.
  """
  def handle_callback!(%Plug.Conn{params: %{"code" => code}} = conn) do
    send_redirect_uri = Keyword.get(options(conn), :send_redirect_uri, true)

    opts =
      if send_redirect_uri do
        [redirect_uri: redirect_uri(conn), code: code]
      else
        [code: code]
      end

    module = option(conn, :oauth2_module)
    token = apply(module, :get_token!, [opts])

    if token.access_token == nil do
      set_errors!(conn, [
        error(token.other_params["error"], token.other_params["error_description"])
      ])
    else
      fetch_user(conn, token)
    end
  end

  @doc """
  Handles a callback from the client when using the Token
  Flow. In this case, the client requests a code, then exchanges it
  for the jwt token. We just need to verify the jwt token with HP ID.
  """
  def handle_callback!(%Plug.Conn{params: %{"access_token" => access_token}} = conn) do
    token = OAuth2.AccessToken.new(access_token)

    # verify that the access token's aud matches our client id
    if check_access_token(conn, token) do
      fetch_user(conn, token)
    else
      set_errors!(conn, [error("token", "Token verification failed")])
    end
  end

  @doc false
  def handle_callback!(conn) do
    set_errors!(conn, [error("missing_code", "No code received")])
  end

  @doc """
  Cleans up the private area of the connection used for passing the raw HP ID response around during the callback.
  """
  def handle_cleanup!(conn) do
    conn
    |> put_private(:hpid_user, nil)
    |> put_private(:hpid_token, nil)
  end

  @doc """
  Fetches the uid field from the HP ID response. This defaults to the option `uid_field` which in-turn defaults to `id`
  """
  def uid(conn) do
    conn |> option(:uid_field) |> to_string() |> fetch_uid(conn)
  end

  @doc """
  Includes the credentials from the HP ID response.
  """
  def credentials(conn) do
    token = conn.private.hpid_token
    scope_string = token.other_params["scope"] || ""
    scopes = String.split(scope_string, ",")

    %Credentials{
      token: token.access_token,
      refresh_token: token.refresh_token,
      expires_at: token.expires_at,
      token_type: token.token_type,
      expires: !!token.expires_at,
      scopes: scopes
    }
  end

  @doc """
  Fetches the fields to populate the info section of the `Ueberauth.Auth` struct.
  """
  def info(conn) do
    user = conn.private.hpid_user

    %Info{
      name: user["name"],
      first_name: user["given_name"],
      last_name: user["family_name"],
      nickname: user["sub"],
      email: user["email"]
    }
  end

  @doc """
  Stores the raw information (including the token) obtained from the HP ID callback.
  """
  def extra(conn) do
    %Extra{
      raw_info: %{
        token: conn.private.hpid_token,
        user: conn.private.hpid_user
      }
    }
  end

  @doc """
  Verify the access token with HP ID.
  It is especially important to verify the
  aud matches our client_id
  """
  def check_access_token(conn, token) do
    module = option(conn, :oauth2_module)
    apply(module, :validate, [token])
  end

  defp fetch_uid("email", %{private: %{hpid_user: user}}) do
    # private email will not be available as :email and must be fetched
    fetch_email!(user)
  end

  defp fetch_uid(field, conn) do
    conn.private.hpid_user[field]
  end

  defp fetch_email!(user) do
    user["email"]
  end

  defp fetch_user(conn, token) do
    conn = put_private(conn, :hpid_token, token)

    case Ueberauth.Strategy.HPID.OAuth.get(token, "/directory/v1/userinfo") do
      {:ok, %OAuth2.Response{status_code: 401, body: _body}} ->
        set_errors!(conn, [error("token", "unauthorized")])

      {:ok, %OAuth2.Response{status_code: status_code, body: user}}
      when status_code in 200..399 ->
        put_private(conn, :hpid_user, user)

      {:error, %OAuth2.Error{reason: reason}} ->
        set_errors!(conn, [error("OAuth2", reason)])
    end
  end

  defp option(conn, key) do
    Keyword.get(options(conn), key, Keyword.get(default_options(), key))
  end

  @doc ~S"""
  Load the dynamic configuration for the redirect_uri
  or fallback to Ueberauth's callback_url.
  """
  def redirect_uri(conn) do
    get_redirect_uri() || callback_url(conn)
  end
  defp get_redirect_uri do
    uri =
      :ueberauth
      |> Application.fetch_env!(Ueberauth.Strategy.HPID.OAuth)
      |> Keyword.get(:redirect_uri)

    cond do
      uri == "" -> nil
      true -> uri
    end
  end
end
